tower-http-util/src/connection.rs

```diff
@@ -0,0 +1,69 @@
+//! Contains all Http Connection utilities.
+
+use futures::{Future, Poll};
+use http::Version;
+use std::net::SocketAddr;
+use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_tcp::TcpStream;
+use tower_service::Service;
+
+/// A Http aware connection creator.
+pub trait HttpMakeConnection<Target>: sealed::Sealed<Target> {
+    /// The transport provided by this service that is HTTP aware.
+    type Connection: HttpConnection + AsyncRead + AsyncWrite;
+
+    /// Errors produced by the connecting service
+    type Error;
+
+    /// The future that eventually produces the transport
+    type Future: Future<Item = Self::Connection, Error = Self::Error>;
+
+    /// Returns `Ready` when it is able to make more connections.
+    fn poll_ready(&mut self) -> Poll<(), Self::Error>;
+
+    /// Connect and return a transport asynchronously
+    fn make_connection(&mut self, target: Target) -> Self::Future;
+}
+
+/// Represents a HTTP aware connection.
+pub trait HttpConnection {
+    /// Returns the version that this stream is set too.
+    fn version(&self) -> Version;
```

Comment:
Is this meant to represent a version determined by ALPN? Maybe it should be `Option<Version>`, then? Since it could be confusing if a connection was made with HTTP2 Prior Knowledge, and yet this says `Version::HTTP_11`.